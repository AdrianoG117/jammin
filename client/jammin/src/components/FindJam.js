import React from 'react';
import Search from './Search'



function FindJam() {

  return (
    <div className="findJam-main">
      <form className="find-form" action="">
      <h1 id="find-jam">ğŸ‘‡ğŸ‘‡ Enter your location to find a jam in your city ğŸ‘‡ğŸ‘‡</h1>
      <Search />
      <button className="find-btn">Search</button>
      </form>
    </div>
  )
}


export default FindJam
