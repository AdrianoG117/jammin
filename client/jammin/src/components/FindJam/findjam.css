.findJam-main {
  height: 90%;
  width: 100vw;
  margin: 1rem auto;
  display: flex;
  flex-direction: column;
}

.findJam-main form {
  width: 100%;
  display: flex;
  flex-direction: column;
}

.findJam-main input {
  padding: 0.3rem;
  height: 60%;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 2rem;
  font-size: 3rem;
}

.find-form {
  display: flex;
  align-items: center;
  height: 20%;
  animation: fromBottom 1s;
}

.jams-list-container {
  margin-top: 2rem;
  display: flex;
  height: 80%;
  width: 100vw;
  margin: 0 auto;
}

.find-btn {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 20px;
  width: 27%;
  height: 25%;
  background-color: rgba(0, 0, 0, 0.7);
  color: white;
  font-size: 1.3rem;
  animation: fromBottom 1.5s;
}

.find-btn:hover {
  background-color: rgba(0, 0, 0, 0.3);
}

#find-jam {
  display: flex;
  justify-content: center;
  align-items: center;
  color: white;
}

.jams-list {
  width: 50%;
  max-height: 85%;
  margin: 2rem;
  overflow-x: scroll;
}

.maps-container {
  width: 50%;
  padding: 3rem;
}

.map-window {
  background-color: black;
  width: 100%;
  height: 100%;
}

.btn-see-event {
  /* background-color: rgba(0, 0, 0, 0.7); */
  padding: 0.5rem;
  color: black;
  border-radius: 20px;
  width: 4rem;
  border: none;
}

.btn-see-event:hover {
  border: 1px solid black;
}

.msg-fail {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: relative;
  width: 30%;
  left: 350px;
  bottom: 250px;
  /* left: 400px; */
}

#search-fail {
  /* background-color: red; */
  color: rgba(0, 0, 0, 0.8);
  width: 80%;
  text-align: center;
}

#btn-fail {
  background-color: rgba(0, 0, 0, 0.7);
  padding: 0.5rem;
  color: white;
  border-radius: 20px;
  width: 30vw;
  height: 10vh;
  margin-top: 2rem;
  font-size: 2.5rem;
}

#btn-fail:hover {
  background-color: rgba(0, 0, 0, 0.3);
}

.error-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: relative;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

@keyframes fromBottom {
  0% {
    transform: translate(0px, 200px);
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
