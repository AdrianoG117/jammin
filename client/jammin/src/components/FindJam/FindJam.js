import React, {useState} from 'react';
import Search from '../Search/Search';
import apiService from '../../ApiService';
import './findjam.css';
import {
  GoogleMap,
  useLoadScript,
  Marker,
  InfoWindow
} from "@react-google-maps/api";
// import { formatRelative } from "adt-fns";

const libraries = ["places"];
const mapContainerStyle = {
  width: "50vw",
  height: "50vh"
};

function FindJam() {

  const [searchVal, setSearchVal] = useState({location: null});
  const [jams, setJams] = useState([]);
  const [center, setCenter] = useState(null);
  const [markers, setMarkers] = useState([]);
  const [errorFail, setErrorFail] = useState("");

  function searchJams (input) {
    console.log('searchJams function running')
    setSearchVal(input)
  }

  function getCoords(input) {
    let result = [];
    for (let i = 0; i < input.length; i++) {
      result.push(input[i].locCords);
    }
    console.log(result);
    return result
  }

  async function handleSubmit (e) {
    e.preventDefault();
    const result = await apiService.getJams({city: searchVal});
    setJams(result);
    console.log(result);
    let eventsCoords = getCoords(result);
    setMarkers(eventsCoords);
    //get coordinates of every element of the array results
    // setMarkers()
    fetch(`https://maps.googleapis.com/maps/api/geocode/json?address=${searchVal}&key=AIzaSyCaWssSgkyqO9SyAJ7VvTonQ1ASzdyQ6oM`)
    .then((res) => res.json())
    .then((data => {
      let coords = data.results[0].geometry.location;
      setCenter(coords)
    }))
  }

  const{ isLoaded, loadError} = useLoadScript({
    googleMapsApiKey: 'AIzaSyCaWssSgkyqO9SyAJ7VvTonQ1ASzdyQ6oM',
    libraries
  })

  if(loadError) return "Error loading maps";
  if (!isLoaded) return "Loading Maps";


  return (
    <div className="findJam-main">
      <form className="find-form" onSubmit={handleSubmit}>
      <h1 id="find-jam">ðŸ‘‡ðŸ‘‡ Enter your location to find a jam in your city ðŸ‘‡ðŸ‘‡</h1>
      <Search searchJams={searchJams}/>
      <button className="find-btn">Search</button>
      <div className="jams-list-container">
        <div className="jams-list">
          {jams.length? jams.map(jam =>
          <div className="single-jam">
            <div className="jam-info">
              <h1>{jam.date}</h1>
              <h1>{jam.title}</h1>
              <div className="jam-loc">
                <img src="" alt=""/>
                <p>{jam.location}</p>
              </div>
            </div>
            <div className="see-event">
              <button>SEE EVENT</button>
            </div>
          </div>
            ): null}
        </div>
        <div className="maps-container">
          <div className="maps">
          {jams.length ?
                    <GoogleMap
                    mapContainerStyle={mapContainerStyle}
                    zoom={13}
                    center={center}
                    >
                      {markers.map(marker => (
                      <Marker
                      position={{ lat: marker.lat, lng: marker.lng}}
                      />
                      ))}
                    </GoogleMap> :
                    <h1 id="search-fail">OOPS it seems there is no jams coming up in this city ðŸ˜…</h1>
                  }
          </div>


        </div>

      </div>
      <div>

      </div>
      </form>
    </div>
  )
}


export default FindJam
