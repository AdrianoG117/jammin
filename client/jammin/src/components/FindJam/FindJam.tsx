import React, { useState, SetStateAction } from "react";
import Search from "../Search/Search";
import JamItem from "../JamItem/JamItem";
import apiService from "../../apiService/ApiService";
import logo from "../../images/marker.png";
import "./findjam.css";
import { Jam } from "../../apiService/APIResponseTypes";
import {
  GoogleMap,
  useLoadScript,
  Marker,
  InfoWindow,
} from "@react-google-maps/api";
import { Link } from "react-router-dom";

const apiKey = process.env.REACT_APP_GOOGLE_MAPS_API_KEY;

const libraries = ["places"];
const mapContainerStyle = {
  width: "100%",
  height: "60vh",
  borderRadius: "10px",
};

interface coords {
    lat: number
    lng: number
  }

interface IProps {
  jams:Jam[] | void;
  setJams: React.Dispatch<SetStateAction< Jam[] >>
  searchVal: string;
  // setSearch:
  setSearchVal: React.Dispatch<SetStateAction<string>>
  center: coords | null;
  setCenter: React.Dispatch<SetStateAction<coords | null>>
  markers: {lat: number, lng: number}[];
  setMarkers: React.Dispatch<SetStateAction<{lat: number, lng: number}[]>>
  hasSearch: boolean;
  setHasSearch: React.Dispatch<SetStateAction<boolean>>
  isSignedUp: boolean
}

const FindJam:React.FunctionComponent<IProps> = ({ jams, setJams, searchVal, setSearchVal, center, setCenter, markers, setMarkers, hasSearch, setHasSearch, isSignedUp }:IProps): React.ReactElement => {

  // const jams = props.jams;
  // console.log(jams);
  // const setJams = props.setJams;
  // const searchVal = props.searchVal;
  // const setSearchVal = props.setSearchVal;
  // const center = props.center;
  // const setCenter = props.setCenter;
  // const markers = props.markers;
  // const setMarkers = props.setMarkers;

  // const hasSearch = props.hasSearch;
  // const setHasSearch = props.setHasSearch;

  // const isSignedUp = props.isSignedUp;

  const [selected, setSelected] = useState<coords | null>(null);
  const [idRoute, setIdRoute] = useState<string | undefined>(undefined);
  const [highEvent, setHighEvent] = useState<coords | null>(null);

  const findPlaceholder = "Enter your city";

  function searchJams(input: string) {
    setSearchVal(input);
  }

  function getCoords(input: Jam[] | void): coords[] | void {
    if (input) {
      let result = [];
      for (let i = 0; i < input.length; i++) {
        result.push(input[i].locCords);
      }
      return result;
    }
  }

  async function handleSubmit(e: any) {
    e.preventDefault();
    const result: void | Jam[] = await apiService.getJams({ city: searchVal });
    setHasSearch(true);

    if (result) setJams(result);
    let eventsCoords = getCoords(result);
    if (eventsCoords) setMarkers(eventsCoords);

    fetch(
      `https://maps.googleapis.com/maps/api/geocode/json?address=${searchVal}&key=${apiKey}`
    )
      .then((res) => res.json())
      .then((data) => {
        let coords = data.results[0].geometry.location;
        setCenter(coords);
      });
  }

  function coordsToId(coords: coords) {
    if (jams)
    for (let i = 0; i < jams.length; i++) {
      if (
        jams[i].locCords.lat === coords.lat &&
        jams[i].locCords.lng === coords.lng
      ) {
        return jams[i]._id;
      }
    }
  }

  const libraries: ("places" | "drawing" | "geometry" | "localContext" | "visualization")[] = ["places"];

  const { isLoaded, loadError } = useLoadScript({
    googleMapsApiKey: apiKey as string,
    libraries,
  });

  return (
    <div className="findJam-main">
      <form className="find-form" onSubmit={handleSubmit}>
        <Search searchJams={searchJams} findPlaceholder={findPlaceholder} />
        <button className="find-btn">Search</button>
      </form>
      <div className="jams-list-container">
        <div className="jams-list">
          {jams?.length
            ? jams.map((jam) => (
                <JamItem
                  jam={jam}
                  setHighEvent={setHighEvent}
                />
              ))
            : null}
        </div>
        <div className="maps-container">
          <div className="maps">
            {jams?.length ? (
              <GoogleMap
                mapContainerStyle={mapContainerStyle}
                zoom={11}
                center={center? center : {lat: 0, lng: 0}}
              >
                {markers.map((marker) => (
                  <Marker
                    onClick={() => {
                      setSelected(marker);
                      const routeId = coordsToId({
                        lat: marker.lat,
                        lng: marker.lng,
                      });
                      setIdRoute(routeId);
                    }}
                    position={{ lat: marker.lat, lng: marker.lng + 0.0034 }}
                    icon={{
                      url: logo,
                      scaledSize:
                        marker.lat === highEvent?.lat &&
                        marker.lng === highEvent?.lng
                          ? new window.google.maps.Size(50, 50)
                          : new window.google.maps.Size(40, 40),
                    }}
                  />
                ))}

                {selected ? (
                  <InfoWindow
                    position={{
                      lat: selected.lat,
                      lng: selected.lng,
                    }}
                    onCloseClick={() => {
                      setSelected(null);
                    }}
                  >
                    <Link to={`/jams/${idRoute}`}>
                      <button className="btn-see-event">SEE EVENT</button>
                      <div className="map-window"></div>
                    </Link>
                  </InfoWindow>
                ) : null}
              </GoogleMap>
            ) : null}
          </div>
        </div>
      </div>
      <div className="error-container">
        {jams?.length === 0 && hasSearch ? (
          <div className="msg-fail">
            <h1 id="search-fail">
              OOPS it seems there is no jams coming up in this city ðŸ˜…
            </h1>
            <Link to="/createjam">
              <button id="btn-fail">Create a jam</button>
            </Link>
          </div>
        ) : null}
      </div>
    </div>
  );
}

export default FindJam;
