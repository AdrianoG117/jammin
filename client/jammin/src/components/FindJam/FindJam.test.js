import {render, screen} from "@testing-library/react";
import FindJam from "./FindJam";

describe("FindJam component", ()=> {
  
    test("should match the snapshot", () => {
      const {container} = render(<FindJam
                    jams={[]} 
                    setJams={()=>{}} 
                    searchVal={""} 
                    setSearchVal={()=>{}} 
                    center={{lat:"", lng:""}} 
                    setCenter={()=>{}} 
                    markers={[]} 
                    setMarkers={()=>{}} 
                    hasSearch={false} 
                    setHasSearch={()=>{}} 
                    isSignedUp={false}
                    />);
      expect(container.firstChild).toMatchSnapshot();
    });
    
    test("Should render search and search button", () => {
      render (<FindJam
                    jams={[]} 
                    setJams={()=>{}} 
                    searchVal={""} 
                    setSearchVal={()=>{}} 
                    center={{lat:"", lng:""}} 
                    setCenter={()=>{}} 
                    markers={[]} 
                    setMarkers={()=>{}} 
                    hasSearch={false} 
                    setHasSearch={()=>{}} 
                    isSignedUp={false} 
                    />);
      //title
      screen.getByPlaceholderText(/Enter your city/);
      screen.getByRole("button", { name: "Search" });
    });

    test("If there are events in your location, display the events and map", () => {

    })

    test("If there are no events in your location, display text and Create a Jam button", () => {

    })

    test("If city is searched display the city in the placeholder", () => {

    })
  
  });