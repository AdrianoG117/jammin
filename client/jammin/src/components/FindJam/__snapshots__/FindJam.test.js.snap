// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindJam component should match the snapshot 1`] = `
<div
  class="findJam-main"
>
  <form
    class="find-form"
  >
    <div
      data-reach-combobox=""
      data-state="idle"
      style="height: 50%; display: flex; align-items: center; justify-content: center;"
    >
      <input
        aria-autocomplete="both"
        aria-controls="listbox--1"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="demo"
        data-reach-combobox-input=""
        data-state="idle"
        disabled=""
        placeholder="Enter your city"
        required=""
        role="combobox"
        style="width: 100%; height: 80%; background-color: rgba(0, 0, 0, 0.7); border-radius: 20px; color: white; padding: 1rem;"
        value=""
      />
    </div>
    <button
      class="find-btn"
    >
      Search
    </button>
  </form>
  <div
    class="jams-list-container"
  >
    <div
      class="jams-list"
    />
    <div
      class="maps-container"
    >
      <div
        class="maps"
      />
    </div>
  </div>
  <div
    class="error-container"
  />
</div>
`;
